### PR TITLE
core/vm: Correct the Memory Gas Overflow condition

### PR DESCRIPTION
**Pull request by LiangMa**
previous overflow condition is too big to use.
0x7FFFFFFFF squre operation is overflowed uint64.

0x7FFFFFFFF^2 = 0x3F FFFF FFF0 0000 0001

the minimal number cause uint64 overflow is 0x100000000^2 = (uint64) 0
test code is 
func main() {
        var min_overflow uint64
        var mem_sz uint64
        var word_sz uint64

        min_overflow = 0x100000000
   
        fmt.Printf("result of 0x%X square op is 0x%X\n", min_overflow, min_overflow*min_overflow)

        mem_sz = 0x1FFFFFFFE0
        word_sz= (mem_sz+31)/32

	fmt.Printf("mem_sz is 0x%X, word sz is 0x%X\n", mem_sz, word_sz)	
}

### CODE REVIEW

**Comment by gballet** [comment]
Please fix gas_table_test to match the new `0x1fffffffe0`value. Also add a test to check values between `0x1fffffffe0` and `0x7fffffffe0`.

It could be optimized by dividing it by 16, checking for the overflow after the sqare, then 2, so as to match 512. But Parity has a similar approach so we don't want to break the consensus here.

**Comment by holiman** [comment]
Thanks. For a second there, I thought you had submitted a real live consensus issue fix. Luckily, that's not quite the case -- even though it overflows, it can only do so on gas limits high enough to be not an issue for mainnet.

**Comment by karalabe** [comment]
@holiman @gballet I've pushed a fix on top of this for the tests + code cleanups. PTAL

File: core/vm/gas_table_test.go [hunk 1]
```diff
@@ -19,18 +19,21 @@ package vm
...
-	if v != 36028899963961341 {
-		t.Errorf("Expected: 36028899963961341, got %d", v)
-	}
-
-	_, err = memoryGasCost(&Memory{}, size+1)
-	if err == nil {
-		t.Error("expected error")
+	for i, tt := range tests {
+		v, err := memoryGasCost(&Memory{}, tt.size)
+		if (err != nil) != tt.overflow {
```
Comment (gballet) [inline core/vm/gas_table_test.go]: @karalabe to be "future-proof", it would be better to check that the error is indeed `errGasUintOverflow`

**Pull request merged**